File: packages/cli/src/commands/history.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import { createApiClient } from "../lib/api-client.js";
import {
  handleError,
  validateDateFormat,
  validatePositiveInt,
} from "../lib/shared.js";

/* ─── remb history ─── */

export const historyCommand = new Command("history")
  .description("View conversation history — what AI discussed and did across sessions")
  .option("-s, --search <query>", "Semantic search across conversation history")
  .option("-d, --date <date>", "Filter by specific date (YYYY-MM-DD)")
  .option("--from <date>", "Start date filter (YYYY-MM-DD)")
  .option("--to <date>", "End date filter (YYYY-MM-DD)")
  .option("-l, --limit <n>", "Max entries to show", "20")
  .option("-p, --project <slug>", "Filter by project slug")
  .option("--format <fmt>", "Output format: timeline (default), markdown, json", "timeline")
  .addHelpText(
    "after",
    `\nExamples:\n  $ remb history\n  $ remb history -d 2025-01-15\n  $ remb history --search "authentication flow"\n  $ remb history --from 2025-01-01 --to 2025-01-31 --format json`,
  )
  .action(async (opts) => {
    // Validate date inputs
    if (opts.date) validateDateFormat(opts.date, "--date");
    if (opts.from) validateDateFormat(opts.from, "--from");
    if (opts.to) validateDateFormat(opts.to, "--to");
    if (opts.from && opts.to && opts.from > opts.to) {
      console.error(chalk.red("\u2716") + " --from date cannot be after --to date.");
      process.exit(1);
    }
    const limit = parseInt(opts.limit, 10) || 20;
    validatePositiveInt(limit, "Limit", 200);

    const spinner = ora("Loading conversation history...").start();

    try {
      const client = createApiClient();
      const projectSlug = opts.project;

      // ── Semantic search mode ──
      if (opts.search) {
        spinner.text = "Searching conversations...";
        const { results } = await client.searchConversations({
          query: opts.search,
          projectSlug,
          limit,
        });
        spinner.stop();

        if (results.length === 0) {
          console.log(chalk.dim("  No matching conversations found."));
          return;
        }

        if (opts.format === "json") {
          console.log(JSON.stringify(results, null, 2));
          return;
        }

        console.log();
        console.log(chalk.bold(`  Search results for "${opts.search}"`));
        console.log(chalk.dim(`  ${results.length} matches\n`));

        for (const r of results) {
          const date = chalk.dim(r.created_at.slice(0, 10));
          const time = chalk.dim(r.created_at.slice(11, 16));
          const sim = chalk.green(`${(r.similarity * 100).toFixed(0)}%`);
          const tags = r.tags?.length ? chalk.blue(` [${r.tags.join(", ")}]`) : "";
          const proj = r.project_slug ? chalk.dim(` (${r.project_slug})`) : "";
          console.log(`  ${date} ${time} ${sim}${proj}${tags}`);
          console.log(`    ${r.content.slice(0, 200)}${r.content.length > 200 ? "..." : ""}`);
          console.log();
        }
        return;
      }

      // Build date params
      let startDate: string | undefined;
      let endDate: string | undefined;
      if (opts.date) {
        startDate = `${opts.date}T00:00:00Z`;
        endDate = `${opts.date}T23:59:59Z`;
      } else {
        if (opts.from) startDate = `${opts.from}T00:00:00Z`;
        if (opts.to) endDate = `${opts.to}T23:59:59Z`;
      }

      // JSON format → raw API response
      if (opts.format === "json") {
        const result = await client.getConversationHistory({
          projectSlug,
          startDate,
          endDate,
          limit,
          format: "json",
        });
        spinner.stop();
        console.log(JSON.stringify(result.entries, null, 2));
        return;
      }

      // Markdown format → server-generated markdown
      if (opts.format === "markdown") {
        const result = await client.getConversationHistory({
          projectSlug,
          startDate,
          endDate,
          limit,
          format: "json",
        });
        spinner.stop();
        printMarkdown(result.entries);
        return;
      }

      // Default: timeline format
      const result = await client.getConversationHistory({
        projectSlug,
        startDate,
        endDate,
        limit,
        format: "json",
      });

      spinner.stop();

      if (result.entries.length === 0) {
        console.log(chalk.dim("  No conversation history. AI sessions log here automatically via MCP."));
        return;
      }

      console.log();
      console.log(chalk.bold("  Conversation History"));
      console.log(chalk.dim(`  ${result.total} entries\n`));

      // Group by date
      const grouped = new Map<string, typeof result.entries>();
      for (const entry of result.entries) {
        const date = entry.created_at.slice(0, 10);
        const list = grouped.get(date) ?? [];
        list.push(entry);
        grouped.set(date, list);
      }

      for (const [date, entries] of grouped) {
        console.log(chalk.bold.blue(`  ${date}`));
        for (const e of entries) {
          const time = chalk.dim(e.created_at.slice(11, 16));
          const icon =
            e.type === "tool_call"
              ? chalk.yellow("⚡")
              : e.type === "milestone"
                ? chalk.green("◆")
                : chalk.cyan("●");
          const src = e.source !== "mcp" ? chalk.dim(` [${e.source}]`) : "";
          console.log(`    ${time} ${icon}${src} ${e.content}`);
        }
        console.log();
      }
    } catch (err) {
      spinner.stop();
      handleError(err);
    }
  });

/* ─── markdown printer ─── */

function printMarkdown(
  entries: Array<{
    created_at: string;
    type: string;
    content: string;
    source: string;
  }>
) {
  if (entries.length === 0) {
    console.log("No conversation history. AI sessions log here automatically via MCP.");
    return;
  }

  // Reverse to chronological
  const chronological = [...entries].reverse();

  const grouped = new Map<string, typeof chronological>();
  for (const entry of chronological) {
    const date = entry.created_at.slice(0, 10);
    const list = grouped.get(date) ?? [];
    list.push(entry);
    grouped.set(date, list);
  }

  console.log("# Conversation History\n");
  for (const [date, dayEntries] of grouped) {
    console.log(`## ${date}\n`);
    for (const e of dayEntries) {
      const time = e.created_at.slice(11, 16);
      const icon = e.type === "tool_call" ? "🔧" : e.type === "milestone" ? "🏁" : "💬";
      const src = e.source !== "mcp" ? ` [${e.source}]` : "";
      console.log(`- **${time}** ${icon}${src} ${e.content}`);
    }
    console.log();
  }
}

File: packages/cli-go/cmd/history.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"

	"github.com/richie/remb/internal/api"
	"github.com/richie/remb/internal/config"
	"github.com/spf13/cobra"
)

var (
	histDate    string
	histFrom    string
	histTo      string
	histLimit   int
	histProject string
	histFormat  string
)

var historyCmd = &cobra.Command{
	Use:   "history",
	Short: "View conversation history — what AI discussed and did across sessions",
	RunE:  runHistory,
}

func init() {
	historyCmd.Flags().StringVarP(&histDate, "date", "d", "", "Filter by specific date (YYYY-MM-DD)")
	historyCmd.Flags().StringVar(&histFrom, "from", "", "Start date filter (YYYY-MM-DD)")
	historyCmd.Flags().StringVar(&histTo, "to", "", "End date filter (YYYY-MM-DD)")
	historyCmd.Flags().IntVarP(&histLimit, "limit", "l", 20, "Max entries to show")
	historyCmd.Flags().StringVarP(&histProject, "project", "p", "", "Filter by project slug")
	historyCmd.Flags().StringVar(&histFormat, "format", "timeline", "Output format: timeline, markdown, json")
}

func runHistory(cmd *cobra.Command, args []string) error {
	client, err := api.NewClient()
	if err != nil {
		return err
	}

	// Resolve project slug
	projectSlug := histProject
	if projectSlug == "" {
		cfg := config.FindProjectConfig("")
		if cfg != nil && cfg.Config.Project != "" {
			projectSlug = cfg.Config.Project
		}
	}

	// Build query params
	params := map[string]string{}
	if projectSlug != "" {
		params["projectSlug"] = projectSlug
	}
	if histDate != "" {
		params["startDate"] = histDate + "T00:00:00Z"
		params["endDate"] = histDate + "T23:59:59Z"
	} else {
		if histFrom != "" {
			params["startDate"] = histFrom + "T00:00:00Z"
		}
		if histTo != "" {
			params["endDate"] = histTo + "T23:59:59Z"
		}
	}
	if histLimit > 0 {
		params["limit"] = fmt.Sprintf("%d", histLimit)
	}

	result, err := client.GetConversationHistory(params)
	if err != nil {
		handleAPIError(err)
		return nil
	}

	switch histFormat {
	case "json":
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		return enc.Encode(result.Entries)

	case "markdown":
		printHistoryMarkdown(result.Entries)
		return nil

	default:
		printHistoryTimeline(result)
		return nil
	}
}

func printHistoryTimeline(result *api.ConversationHistoryResponse) {
	if len(result.Entries) == 0 {
		fmt.Fprintf(os.Stderr, "  %sNo conversation history found.%s\n", "\033[2m", "\033[0m")
		return
	}

	fmt.Fprintf(os.Stderr, "\n  %sConversation History%s\n", "\033[1m", "\033[0m")
	fmt.Fprintf(os.Stderr, "  %s%d entries%s\n\n", "\033[2m", result.Total, "\033[0m")

	// Group by date
	grouped := groupByDate(result.Entries)
	dates := sortedKeys(grouped)

	for _, date := range dates {
		entries := grouped[date]
		fmt.Fprintf(os.Stderr, "  \033[1;34m%s\033[0m\n", date)
		for _, e := range entries {
			time := e.CreatedAt[11:16]
			icon := "\033[36m●\033[0m" // cyan dot = summary
			if e.Type == "tool_call" {
				icon = "\033[33m⚡\033[0m" // yellow bolt = tool call
			} else if e.Type == "milestone" {
				icon = "\033[32m◆\033[0m" // green diamond = milestone
			}
			src := ""
			if e.Source != "mcp" {
				src = fmt.Sprintf(" \033[2m[%s]\033[0m", e.Source)
			}
			fmt.Fprintf(os.Stderr, "    \033[2m%s\033[0m %s%s %s\n", time, icon, src, e.Content)
		}
		fmt.Fprintln(os.Stderr)
	}
}

func printHistoryMarkdown(entries []api.ConversationEntry) {
	if len(entries) == 0 {
		fmt.Println("No conversation history found.")
		return
	}

	// Reverse to chronological
	reversed := make([]api.ConversationEntry, len(entries))
	copy(reversed, entries)
	sort.Slice(reversed, func(i, j int) bool {
		return reversed[i].CreatedAt < reversed[j].CreatedAt
	})

	grouped := groupByDate(reversed)
	dates := sortedKeys(grouped)

	fmt.Println("# Conversation History")
	fmt.Println()
	for _, date := range dates {
		dayEntries := grouped[date]
		fmt.Printf("## %s\n\n", date)
		for _, e := range dayEntries {
			time := e.CreatedAt[11:16]
			icon := "💬"
			if e.Type == "tool_call" {
				icon = "🔧"
			} else if e.Type == "milestone" {
				icon = "🏁"
			}
			src := ""
			if e.Source != "mcp" {
				src = fmt.Sprintf(" [%s]", e.Source)
			}
			fmt.Printf("- **%s** %s%s %s\n", time, icon, src, e.Content)
		}
		fmt.Println()
	}
}

func groupByDate(entries []api.ConversationEntry) map[string][]api.ConversationEntry {
	grouped := map[string][]api.ConversationEntry{}
	for _, e := range entries {
		date := e.CreatedAt[:10]
		grouped[date] = append(grouped[date], e)
	}
	return grouped
}

func sortedKeys(m map[string][]api.ConversationEntry) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}
